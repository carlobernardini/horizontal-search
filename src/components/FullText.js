import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { debounce, isFunction } from 'lodash-es';
import { DISPLAY_NAME_FILTER_FULLTEXT } from '../constants';
import Caption from './Caption';
import FilterInfo from './FilterInfo';

class FullText extends Component {
    constructor(props) {
        super(props);
        this.state = {
            derivedValue: props.value,
            value: props.value || ''
        };

        this.dispatchDebounced = debounce(
            this.dispatchToQuery,
            props.dispatchTimeout
        );
        this.input = React.createRef();
        this.dispatchDebouncedWrapper = this.dispatchDebouncedWrapper.bind(this);
        this.dispatchToQuery = this.dispatchToQuery.bind(this);
    }

    componentDidMount() {
        const { autoFocus } = this.props;

        if (!autoFocus || !this.input || !this.input.current) {
            return false;
        }

        this.input.current.focus();
        return true;
    }

    static getDerivedStateFromProps(nextProps, prevState) {
        const { value } = nextProps;
        const { derivedValue } = prevState;

        if (value === null && derivedValue !== value) {
            return {
                derivedValue: value,
                value: ''
            };
        }

        return {
            derivedValue: value
        };
    }

    dispatchDebouncedWrapper() {
        if (!this.dispatchDebounced) {
            return true;
        }
        if (this.dispatchDebounced.cancel) {
            this.dispatchDebounced.cancel();
        }
        this.dispatchDebounced();
        return true;
    }

    dispatchToQuery() {
        const { onUpdate, validateInput } = this.props;
        const { value = '' } = this.state;

        if (isFunction(validateInput) && !validateInput(value)) {
            return false;
        }

        if (!value.trim().length) {
            return onUpdate(null);
        }

        return onUpdate(value);
    }

    renderField() {
        const {
            disabled, maxHeight, maxWidth, multiline, validateInput,
            placeholder, value: derivedValue
        } = this.props;

        const { value = '' } = this.state;

        let fieldProps = {
            className: classNames('orizzonte__filter-fulltext', {
                'orizzonte__filter-fulltext--disabled': disabled,
                'orizzonte__filter-fulltext--invalid': value.length && isFunction(validateInput) && !validateInput(value)
            }),
            disabled,
            onBlur: this.dispatchToQuery,
            onChange: (e) => {
                const { value: val = '' } = e.target;

                if (val === value) {
                    return false;
                }

                this.setState({
                    derivedValue,
                    value: val
                }, this.dispatchDebouncedWrapper);
                return true;
            },
            placeholder,
            ref: this.input,
            value
        };

        if (multiline) {
            fieldProps = {
                ...fieldProps,
                className: classNames(fieldProps.className, 'orizzonte__filter-fulltext--multiline'),
                style: ((mh, mw) => {
                    if (!mh && !mw) {
                        return null;
                    }

                    return {
                        maxHeight: mh ? `${ mh }px` : null,
                        maxWidth: mw ? `${ mw }px` : null
                    };
                })(maxHeight, maxWidth)
            };

            return (
                <textarea
                    { ...fieldProps }
                />
            );
        }

        return (
            <input
                type="text"
                onKeyUp={ (e) => {
                    const key = e.keyCode || e.which;

                    if (key !== 13) {
                        return true;
                    }

                    this.dispatchToQuery();
                    return true;
                }}
                { ...fieldProps }
            />
        );
    }

    render() {
        const { information, label } = this.props;

        return (
            <div
                className="orizzonte__filter"
            >
                <FilterInfo
                    information={ information }
                />
                <Caption>
                    { label }
                </Caption>
                { this.renderField() }
            </div>
        );
    }
}

FullText.displayName = DISPLAY_NAME_FILTER_FULLTEXT;

FullText.propTypes = {
    /** If the textarea should focus when group is expanded
        Only works when group has only one (enabled) fulltext filter */
    autoFocus: PropTypes.bool,
    /** If the textarea should be disabled */
    disabled: PropTypes.bool,
    /** Custom debounce timeout before dispatching the new value to the query object */
    dispatchTimeout: PropTypes.number,
    /** Field name for this filter, to be used in composed query */
    // eslint-disable-next-line
    fieldName: PropTypes.string,
    information: PropTypes.string,
    /** Label for this filter section */
    label: PropTypes.string,
    /** Maximum textarea height (only applicable in multiline mode) */
    maxHeight: PropTypes.number,
    /** Maximum textarea width (only applicable in multiline mode) */
    maxWidth: PropTypes.number,
    /** Whether to render a textarea (true) or input field (false) */
    multiline: PropTypes.bool,
    /** Internal callback for filter update */
    onUpdate: PropTypes.func,
    /** Placeholder text for the input field */
    placeholder: PropTypes.string,
    /** Transforming function or placeholder for group label */
    selectedLabel: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.string
    ]),
    /** Function to validate user input, should return true (valid) or false (invalid) */
    validateInput: PropTypes.func,
    /** Current value for this filter */
    value: PropTypes.string
};

FullText.defaultProps = {
    autoFocus: false,
    disabled: false,
    dispatchTimeout: 300,
    fieldName: null,
    information: null,
    label: null,
    maxHeight: null,
    maxWidth: null,
    multiline: false,
    onUpdate: () => {},
    placeholder: null,
    selectedLabel: '%s',
    validateInput: null,
    value: null
};

export default FullText;
