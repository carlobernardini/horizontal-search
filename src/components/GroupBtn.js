import React from 'react';
import PropTypes from 'prop-types';
import { DISPLAY_NAME_GROUP_BTN } from '../constants';

const GroupBtn = ({ hidden, onClick }) => {
    if (hidden) {
        return null;
    }

    return (
        <button
            type="button"
            className="orizzonte__group-btn"
            onClick={ onClick }
        >
            &nbsp;
        </button>
    );
};

GroupBtn.displayName = DISPLAY_NAME_GROUP_BTN;

GroupBtn.propTypes = {
    hidden: PropTypes.bool,
    onClick: PropTypes.func
};

GroupBtn.defaultProps = {
    hidden: false,
    onClick: () => {}
};

export default GroupBtn;
