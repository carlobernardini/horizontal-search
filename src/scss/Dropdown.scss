@import 'base';

.orizzonte__dropdown {
	position: relative;

	&--focused {
		.orizzonte__dropdown-button,
		.orizzonte__dropdown-filter {
			border-color: $color-a;
		}
	}
	&--expanded {
		.orizzonte__dropdown-list {
			display: block;
		}
	}

	&--disabled {
		.orizzonte__dropdown-button {
	        @extend %select-disabled;
	    }
    }
}

.orizzonte__dropdown-button {
	@extend %select;
	text-align: left;
}

.orizzonte__dropdown-filter {
	@extend %select;
	background-image: none;
	box-sizing: border-box;
	padding-right: 31px;

	&-wrapper {
		position: relative;
	}

	&-button,
	&-loading {
		position: absolute;
		right: 1px;
		top: 1px;
		width: 29px;
		height: 30px;
		padding: 0;
		border: none;
		outline: none;
        border-left: 1px solid $border-color-inside;
	}

	&-button {
		background: transparent url($caret-base64);
	    background-repeat: no-repeat;
	    background-position: 50% 50%;
	    background-size: 8px auto;
	}

    &-loading {
        width: 28px;
        .orizzonte__loading {
            position: absolute;
            top: 8px;
            right: 9px;
        }
    }
}

.orizzonte__dropdown-list {
	@extend %list;
	display: none;
	border-radius: 3px;
	margin-top: 3px;
	min-width: 100%;
	max-height: 150px;
	overflow: auto;
	overflow-x: hidden;
	z-index: 999;
	font: {
		size: .9em;
	}
	animation: dropdownListIn .1s linear;
}

.orizzonte__dropdown-item {
	@extend %list-item;
	padding: {
		top: 5px;
		right: 6px;
		left: 6px;
		bottom: 5px;
	}

	.orizzonte__checkbox {
		margin: {
			top: -5px;
			right: -6px;
			bottom: -5px;
			left: -6px;
		}
	}

	.orizzonte__checkbox-label {
		padding: {
			top: 5px;
			right: 6px;
			bottom: 5px;
			left: 6px;
		}
	}

	.orizzonte__checkbox-span {
		&--first {
			width: 9px;
        	height: 9px;
		}
		&--last {
			padding: 0 0 0 5px;
		}
	}

	&:hover {
		.orizzonte__checkbox-span {
			&--first {
				border-color: #fff;
				background-color:#fff;
			}
		    &--last {
		        color: #fff;
		    }
		}
	}

	&--empty {
		color: $text-color-muted;
		line-height: 16px;
		padding: {
			top: 5px;
			right: 6px;
			left: 6px;
			bottom: 5px;
		}
	}

	&--disabled {
		cursor: not-allowed;
		color: $text-color-muted;

		&:hover {
			background-color: transparent;
			color: $text-color-muted;

			.orizzonte__checkbox-span {
				&--first {
					border-color: $border-color;
				}
			    &--last {
			        color: $text-color-muted;
			    }
			}
		}
	}
}

.orizzonte__dropdown-group {
	padding-left: 0;

	.orizzonte__dropdown-item {
		padding-left: 22px;
	}

	&-label {
		background-color: $background-color-muted;
		color: $text-color;
	}
}

.orizzonte__dropdown-match {
	font-weight: 500;
}
