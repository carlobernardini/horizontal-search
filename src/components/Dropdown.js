import React, { Component } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import classNames from 'classnames';
import {
    assign, debounce, filter as _filter, includes, indexOf,
    isEqual, isFunction, unionBy, uniqueId, without
} from 'lodash';
import diacritics from 'diacritics';
import CheckBox from './CheckBox';
import LoadingIndicator from './LoadingIndicator';
import '../scss/Dropdown.scss';

class Dropdown extends Component {
    constructor(props) {
        super(props);

        this.state = {
            expanded: false,
            filter: null,
            focused: false,
            remoteLoading: false,
            remoteOptions: [] // Options that were fetched from a remote API
        };

        this.debounceRemote = null;
        this.dropdown = React.createRef();
        this.filter = React.createRef();
        this.toggleDropdown = this.toggleDropdown.bind(this);
        this.onFocus = this.onFocus.bind(this);
        this.onBlur = this.onBlur.bind(this);
        this.handleClickOutside = this.handleClickOutside.bind(this);
        this.handleEscPress = this.handleEscPress.bind(this);
        this.queryRemote = this.queryRemote.bind(this);
    }

    componentDidUpdate() {
        if (this.filter && this.filter.current) {
            this.filter.current.focus();
        }
    }

    onFocus() {
        const { focused } = this.state;

        if (focused) {
            return false;
        }

        this.setState({
            focused: true
        });
        return true;
    }

    onBlur() {
        const { focused } = this.state;

        if (!focused) {
            return false;
        }

        this.setState({
            focused: false
        });
        return true;
    }

    getFilteredOptions() {
        const { filter } = this.state;

        const mergedOptions = this.getMergedOptions();

        if (!filter) {
            return mergedOptions;
        }

        const re = new RegExp(`(${ filter })`, 'gi');

        return mergedOptions.filter((option) => {
            const label = option.label || option.value;
            return label.match(re);
        });
    }

    getHighlightedLabel(label) {
        const { filter } = this.state;

        if (!filter) {
            return label;
        }

        const parts = label.split(new RegExp(`(${ filter })`, 'gi'));

        return (
            <span>
                { parts.map((part, i) => {
                    if (part.toLowerCase() === filter.toLowerCase()) {
                        return (
                            <strong
                                key={ i }
                                className="orizzonte__dropdown-match"
                            >
                                { part }
                            </strong>
                        );
                    }
                    return (
                        <span key={ i }>
                            { part }
                        </span>
                    );
                }) }
            </span>
        );
    }

    getMergedOptions() {
        const { options } = this.props;
        const { remoteOptions } = this.state;

        return unionBy(options, remoteOptions, 'value');
    }

    handleClickOutside(e) {
        if (!this.dropdown || !this.dropdown.current) {
            return false;
        }
        if (this.dropdown.current.contains(e.target)) {
            return false;
        }
        this.toggleDropdown(null, true, true);
        return true;
    }

    handleEscPress(e) {
        e.stopPropagation();
        const key = e.keyCode || e.which;

        if (key !== 27) {
            return false;
        }

        this.toggleDropdown(null, true, true);
        return false;
    }

    toggleDropdown(e, collapse = false, focusOut = false) {
        const { disabled } = this.props;
        const { expanded, filter } = this.state;

        const newState = {
            expanded: collapse ? false : !expanded
        };

        if (newState.expanded && disabled) {
            return false;
        }
        if (!newState.expanded && filter) {
            newState.filter = null;
        }
        if (focusOut) {
            newState.focused = false;
        }

        if (newState.expanded) {
            document.addEventListener('click', this.handleClickOutside, false);
            document.addEventListener('keyup', this.handleEscPress, true);
        } else {
            document.removeEventListener('click', this.handleClickOutside, false);
            document.removeEventListener('keyup', this.handleEscPress, true);
        }

        this.setState(newState, this.queryRemote);

        return true;
    }

    queryRemote() {
        const { remote } = this.props;
        const { expanded, filter, remoteOptions } = this.state;

        if (!expanded || !remote || !remote.endpoint || !remote.searchParam) {
            return false;
        }

        this.setState({
            remoteLoading: true
        }, () => {
            axios
                .get(remote.endpoint, {
                    data: assign({}, remote.data || {}, {
                        [remote.searchParam]: filter || ''
                    })
                })
                .then((response) => {
                    let { data } = response;
                    const newState = {
                        remoteLoading: false
                    };

                    if (remote.transformer && isFunction(remote.transformer)) {
                        // Use a custom callback to transform the remote response
                        // so the result conforms to the expected data structure
                        // (collection of options)
                        data = remote.transformer(data);
                    }

                    if (!isEqual(data, remoteOptions)) {
                        newState.remoteOptions = data;
                    }

                    this.setState(newState);
                });
        });

        return true;
    }

    renderButtonLabel() {
        const { notSetLabel, value, selectedLabel } = this.props;
        const mergedOptions = this.getMergedOptions();

        if (!value || !value.length) {
            return notSetLabel || 'None selected';
        }

        const selectedOptions = _filter(mergedOptions, (option) => {
            if (Array.isArray(value)) {
                return indexOf(value, option.value) > -1;
            }
            return option.value === value;
        });

        if (!selectedLabel) {
            if (selectedOptions.length === 1) {
                return selectedOptions[0].label || value;
            }
            return `${ selectedOptions.length } selected`;
        }
        if (isFunction(selectedLabel)) {
            return selectedLabel(selectedOptions);
        }
        if (indexOf(selectedLabel, '%d') > -1) {
            return selectedLabel.replace('%d', selectedoptions.length);
        }
        return selectedLabel.replace('%s', selectedOptions[0].label || value);
    }

    renderDropdownTriggerButton() {
        const { remoteLoading } = this.state;

        if (remoteLoading) {
            return (
                <div
                    className="orizzonte__dropdown-filter-loading"
                >
                    <LoadingIndicator
                        size="12"
                    />
                </div>
            );
        }

        return (
            <button
                className="orizzonte__dropdown-filter-button"
                onClick={ () => (this.toggleDropdown(null, true, true)) }
                type="button"
            >
                &nbsp;
            </button>
        );
    }

    renderDropdownTrigger() {
        const {
            disabled, filter
        } = this.props;
        const { expanded } = this.state;

        if (filter && filter.enabled && expanded) {
            return (
                <div className="orizzonte__dropdown-filter-wrapper">
                    <input
                        type="text"
                        className="orizzonte__dropdown-filter"
                        onChange={ (e) => {
                            const { value } = e.target;
                            this.setState({
                                filter: filter.matchDiacritics ? value : diacritics.remove(value)
                            }, () => {
                                if (!this.debounceRemote) {
                                    this.debounceRemote = debounce(this.queryRemote, 300);
                                } else {
                                    this.debounceRemote.cancel();
                                }
                                this.debounceRemote();
                            });
                        }}
                        placeholder={ filter.placeholder || '' }
                        ref={ this.filter }
                    />
                    { this.renderDropdownTriggerButton() }
                </div>
            );
        }

        return (
            <button
                className="orizzonte__dropdown-button"
                disabled={ disabled }
                onClick={ this.toggleDropdown }
                onFocus={ this.onFocus }
                onBlur={ this.onBlur }
                type="button"
            >
                { this.renderButtonLabel() }
            </button>
        );
    }

    renderItem(option) {
        const { multiple, onUpdate, value } = this.props;

        const highlightedLabel = this.getHighlightedLabel(option.label || option.value);

        if (!multiple) {
            return highlightedLabel;
        }

        return (
            <CheckBox
                disabled={ option.disabled }
                id={ uniqueId('checkbox-') }
                value={ option.value }
                label={ highlightedLabel }
                selected={ (value || []).indexOf(option.value) > -1 }
                onChange={ (selected) => {
                    let newValue = (value || []).slice(0);
                    if (selected && !includes(newValue, option.value)) {
                        newValue.push(option.value);
                    }
                    if (!selected && includes(newValue, option.value)) {
                        newValue = without(newValue, option.value);
                    }
                    if (isEqual(newValue, value)) {
                        return false;
                    }
                    onUpdate(newValue.length ? newValue : null);
                    return true;
                }}
                viewBox={[0, 0, 13, 13]}
            />
        );
    }

    renderList() {
        const { filter, remoteLoading } = this.state;
        const options = this.getFilteredOptions();

        if (!options.length) {
            const noOptionsLabel = filter ? 'No matches' : 'No options available';

            return (
                <li
                    className="orizzonte__dropdown-item--empty"
                >
                    { remoteLoading ? 'Loading...' : noOptionsLabel }
                </li>
            );
        }

        return options.map((option, i) => (
            <li
                key={ `${ option.value }.${ i }` }
                className="orizzonte__dropdown-item"
            >
                { this.renderItem(option) }
            </li>
        ));
    }

    render() {
        const { disabled, label } = this.props;
        const { expanded, focused } = this.state;

        return (
            <div
                className="orizzonte__filter"
            >
                <div
                    className="orizzonte__filter-caption"
                >
                    { label }
                </div>
                <div
                    className={ classNames('orizzonte__dropdown', {
                        'orizzonte__dropdown--focused': expanded || focused,
                        'orizzonte__dropdown--expanded': expanded,
                        'orizzonte__dropdown--disabled': disabled
                    }) }
                    ref={ this.dropdown }
                >
                    { this.renderDropdownTrigger() }
                    <ul
                        className="orizzonte__dropdown-list"
                    >
                        { this.renderList() }
                    </ul>
                </div>
            </div>
        );
    }
}

Dropdown.propTypes = {
    disabled: PropTypes.bool,
    /** Filter dropdown options and highlight matches */
    filter: PropTypes.shape({
        enabled: PropTypes.bool.isRequired,
        matchDiacritics: PropTypes.bool,
        placeholder: PropTypes.string
    }),
    label: PropTypes.string.isRequired,
    multiple: PropTypes.bool,
    /** Label to shown when no options are selected */
    notSetLabel: PropTypes.string,
    onUpdate: PropTypes.func,
    /** Collection of dropdown options */
    options: PropTypes.arrayOf(
        PropTypes.shape({
            value: PropTypes.oneOfType([
                PropTypes.number,
                PropTypes.string
            ]).isRequired,
            label: PropTypes.any
        })
    ),
    /** Remote API to fetch dropdown options from */
    remote: PropTypes.shape({
        data: PropTypes.object,
        endpoint: PropTypes.string.isRequired,
        searchParam: PropTypes.string.isRequired,
        transformer: PropTypes.func
    }),
    selectedLabel: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.func
    ]),
    value: PropTypes.array
};

Dropdown.defaultProps = {
    disabled: false,
    filter: null,
    multiple: true,
    notSetLabel: null,
    onUpdate: () => {},
    options: [],
    remote: null,
    selectedLabel: null,
    value: []
};

export default Dropdown;
