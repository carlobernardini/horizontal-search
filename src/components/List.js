import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import '../scss/List.scss';

class List extends Component {
    renderDoneBtn() {
        const {
            doneBtn, doneBtnLabel, isFilterGroup, onApply
        } = this.props;

        if (!isFilterGroup || !doneBtn) {
            return null;
        }

        return (
            <li>
                <button
                    type="button"
                    className="orizzonte__list-done"
                    onClick={ onApply }
                >
                    { doneBtnLabel || 'Done' }
                </button>
            </li>
        );
    }

    renderItems() {
        const {
            values, items, isFilterGroup, onUpdate
        } = this.props;

        if (isFilterGroup) {
            return React.Children.map(items, (item, i) => (
                <li
                    className={ classNames('orizzonte__item', {
                        'orizzonte__item--filters': isFilterGroup
                    }) }
                    key={ i }
                >
                    { React.cloneElement(item, {
                        value: ((v, fn) => {
                            if (!(fn in v)) {
                                return null;
                            }
                            return v[fn];
                        })(values, item.props.fieldName),
                        onUpdate: (filterValue) => {
                            const { fieldName } = item.props;
                            onUpdate(fieldName, filterValue);
                        }
                    }) }
                </li>
            ));
        }

        return items.map((item, i) => (
            <li
                className={ classNames('orizzonte__item', {
                    'orizzonte__item--filters': isFilterGroup
                }) }
                key={ i }
            >
                { item }
            </li>
        ));
    }

    render() {
        const { orientation } = this.props;

        return (
            <ul
                className={ classNames('orizzonte__list', {
                    'orizzonte__list--right': orientation === 'right'
                }) }
            >
                { this.renderItems() }
                { this.renderDoneBtn() }
            </ul>
        );
    }
}

List.propTypes = {
    doneBtn: PropTypes.bool,
    doneBtnLabel: PropTypes.string,
    values: PropTypes.object,
    isFilterGroup: PropTypes.bool,
    items: PropTypes.array.isRequired,
    onApply: PropTypes.func,
    onUpdate: PropTypes.func,
    orientation: PropTypes.oneOf([
        'left',
        'right'
    ])
};

List.defaultProps = {
    doneBtn: true,
    doneBtnLabel: null,
    values: {},
    isFilterGroup: false,
    onApply: () => {},
    onUpdate: () => {},
    orientation: 'left'
};

export default List;
