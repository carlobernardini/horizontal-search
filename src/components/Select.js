import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import '../scss/Filter.scss';
import '../scss/Select.scss';

const Select = ({ disabled, label, onUpdate, options, value }) => (
    <div
        className="orizzonte__filter"
    >
        <div
            className="orizzonte__filter-caption"
        >
            { label }
        </div>
        <select
            className={ classNames('orizzonte__filter-select', {
                'orizzonte__filter-select--disabled': disabled
            }) }
            disabled={ disabled }
            onChange={ (e) => {
                const { value: val } = e.target;
                onUpdate(val);
            }}
            value={ value }
        >
            { options.map((option, i) => (
                <option
                    key={ `${ option.value }.${ i }` }
                    value={ option.value }
                >
                    { option.label || option.value }
                </option>
            )) }
        </select>
    </div>
);

Select.propTypes = {
    /** If the select should be disabled */
    disabled: PropTypes.bool,
    /** Label for this filter section */
    label: PropTypes.string.isRequired,
    /** Internal callback for when select value has changed */
    onUpdate: PropTypes.func,
    /** List of selectable options (value is required) */
    options: PropTypes.arrayOf(
        PropTypes.shape({
            label: PropTypes.string,
            value: PropTypes.oneOfType([
                PropTypes.number,
                PropTypes.string
            ]).isRequired
        })
    ).isRequired,
    value: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string
    ])
};

Select.defaultProps = {
    disabled: false,
    onUpdate: () => {},
    value: null
};

export default Select;
