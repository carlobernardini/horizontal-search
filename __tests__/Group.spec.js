import React from 'react';
import Dropdown from '../src/components/Dropdown';
import FullText from '../src/components/FullText';
import Group from '../src/components/Group';
import GroupBtn from '../src/components/GroupBtn';
import List from '../src/components/List';
import Select from '../src/components/Select';

describe('<Group />', () => {
    jest.useFakeTimers();

    it('should render a basic group', () => {
        const queryPart = {
            testSelect: 1
        };
        const onGroupToggle = jest.fn();
        const onGroupRemove = jest.fn();
        const onUpdate = jest.fn();

        const wrapper = shallow(
            <Group
                i={ 1 }
                label="Test group"
                onGroupToggle={ onGroupToggle }
                onGroupRemove={ onGroupRemove }
                onUpdate={ onUpdate }
                queryPart={ queryPart }
                groupTopLabels
                included
            >
                <Select
                    fieldName="testSelect"
                    label="Test select"
                    options={ [{
                        label: 'Test value 1',
                        value: 1
                    }, {
                        label: 'Test value 2',
                        value: 2
                    }] }
                    selectedLabel="testSelect: %s"
                />
            </Group>
        );

        expect(wrapper).toMatchSnapshot();

        expect(wrapper.instance().renderLabel()).toBe('testSelect: Test value 1');

        wrapper.find('.orizzonte__group-label').simulate('click');
        expect(onGroupToggle).toHaveBeenCalledWith(1);

        wrapper.setProps({
            active: true
        });

        wrapper.find(List).first().prop('onRemove')();
        expect(wrapper.state().removing).toBe(true);

        expect(onUpdate).toHaveBeenCalledWith(Object.keys(queryPart));
        jest.advanceTimersByTime(500);
        expect(onGroupRemove).toHaveBeenCalledWith(1);
    });

    it('should render an active group with description and min width', () => {
        const map = {};

        document.addEventListener = jest.fn((event, callback) => {
            map[event] = callback;
        });

        const onGroupToggle = jest.fn();

        const wrapper = shallow(
            <Group
                label="Test group"
                description="Description for this group"
                onGroupToggle={ onGroupToggle }
                groupTopLabels
                included
                active
            >
                <Select
                    fieldName="testSelect"
                    label="Test select"
                    options={ [{
                        label: 'Test value 1',
                        value: 1
                    }, {
                        label: 'Test value 2',
                        value: 2
                    }] }
                />
            </Group>
        );

        expect(wrapper).toMatchSnapshot();

        map.keyup({
            keyCode: 27
        });

        expect(onGroupToggle).toHaveBeenCalledWith(false);
    });

    it('should render group without clear / remove / done buttons, and higher min width', () => {
        const wrapper = shallow(
            <Group
                label="Test group"
                listMinWidth={ 400 }
                queryPart={{
                    testSelect: 1
                }}
                groupTopLabels
                included
                active
                hideClear
                hideRemove
                hideDone
            >
                <Select
                    fieldName="testSelect"
                    label="Test select"
                    options={ [{
                        label: 'Test value 1',
                        value: 1
                    }, {
                        label: 'Test value 2',
                        value: 2
                    }] }
                />
            </Group>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render an active group without top label and no remove button', () => {
        const wrapper = shallow(
            <Group
                label="Test group"
                description="Description for this group"
                hideRemove
                included
                active
            >
                <Select
                    fieldName="testSelect"
                    label="Test select"
                    options={ [{
                        label: 'Test value 1',
                        value: 1
                    }, {
                        label: 'Test value 2',
                        value: 2
                    }] }
                />
            </Group>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render nothing when the group is not included', () => {
        const wrapper = shallow(
            <Group
                label="Test group"
            >
                <Select
                    fieldName="testSelect"
                    label="Test select"
                    options={ [{
                        label: 'Test value 1',
                        value: 1
                    }, {
                        label: 'Test value 2',
                        value: 2
                    }] }
                />
            </Group>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should return a correct label for multiselects', () => {
        const wrapper = shallow(
            <Group
                queryPart={{
                    testDropdown: [1, 2]
                }}
                label="Some group"
            >
                <Dropdown
                    fieldName="testDropdown"
                    label="My dropdown"
                    options={ [{
                        label: 'First',
                        value: 1
                    }, {
                        label: 'Second',
                        value: 2
                    }, {
                        label: 'Third',
                        value: 3
                    }] }
                    selectedLabel={ (options) => {
                        const labels = options.map((option) => (option.label.toLowerCase()));
                        return `Selected ${ labels.join(', ') }`;
                    }}
                />
            </Group>
        );

        expect(wrapper.instance().renderLabel()).toBe('Selected first, second');
    });

    it('should render a correct label for filters without options (e.g. fulltext)', () => {
        const wrapper = shallow(
            <Group
                queryPart={{
                    testFullText: 'test'
                }}
                label="Some group"
            >
                <FullText
                    fieldName="testFullText"
                    label="My fulltext filter"
                    selectedLabel="Value equals %s"
                />
            </Group>
        );

        expect(wrapper.instance().renderLabel()).toBe('Value equals test');
    });

    it('should not render top labels if no filters available and test error boundary', () => {
        const wrapper = shallow(
            <Group
                label="Test group"
                groupTopLabels
                included
            />
        );

        expect(wrapper.instance().renderTopLabel()).toBeNull();

        wrapper.setState({
            hasError: true
        });

        expect(wrapper).toMatchSnapshot();
    })
});
