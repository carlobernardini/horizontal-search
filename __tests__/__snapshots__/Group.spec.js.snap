// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Group /> should render a basic group 1`] = `
<div
  className="orizzonte__group orizzonte__group--clearable"
  style={
    Object {
      "minWidth": "30px",
    }
  }
>
  <ForwardRef
    shown={true}
  >
    Test group
  </ForwardRef>
  <div
    className="orizzonte__group-btn-wrapper"
  >
    <button
      className="orizzonte__group-label"
      onClick={[Function]}
      type="button"
    >
      Test group
    </button>
    <OrizzonteGroupBtn
      hidden={false}
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`<Group /> should render an active group with description and min width 1`] = `
<div
  className="orizzonte__group orizzonte__group--shown orizzonte__group--empty"
  style={
    Object {
      "minWidth": "30px",
    }
  }
>
  <ForwardRef
    shown={true}
  >
    Test group
  </ForwardRef>
  <div
    className="orizzonte__group-btn-wrapper"
  >
    <button
      className="orizzonte__group-label"
      onClick={[Function]}
      type="button"
    >
      Test group
    </button>
    <OrizzonteGroupBtn
      hidden={true}
      onClick={[Function]}
    />
  </div>
  <OrizzonteList
    cache={Object {}}
    doneBtn={true}
    doneBtnLabel={null}
    isFilterGroup={true}
    minWidth={null}
    onApply={[Function]}
    onRemove={[Function]}
    onUpdate={[Function]}
    orientation="left"
    removeBtn={true}
    removeBtnLabel={null}
    syncCacheToGroup={[Function]}
    values={Object {}}
  >
    <div
      className="orizzonte__group-description"
      key="orizzonte-description"
    >
      Description for this group
    </div>
    <OrizzonteSelect
      disabled={false}
      fieldName="testSelect"
      information={null}
      label="Test select"
      notSetLabel={false}
      onUpdate={[Function]}
      options={
        Array [
          Object {
            "label": "Test value 1",
            "value": 1,
          },
          Object {
            "label": "Test value 2",
            "value": 2,
          },
        ]
      }
      value=""
    />
  </OrizzonteList>
</div>
`;

exports[`<Group /> should render an active group without top label and no remove button 1`] = `
<div
  className="orizzonte__group orizzonte__group--shown orizzonte__group--empty"
  style={
    Object {
      "minWidth": "30px",
    }
  }
>
  <div
    className="orizzonte__group-btn-wrapper"
  >
    <button
      className="orizzonte__group-label"
      onClick={[Function]}
      type="button"
    >
      Test group
    </button>
    <OrizzonteGroupBtn
      hidden={true}
      onClick={[Function]}
    />
  </div>
  <OrizzonteList
    cache={Object {}}
    doneBtn={true}
    doneBtnLabel={null}
    isFilterGroup={true}
    minWidth={null}
    onApply={[Function]}
    onRemove={[Function]}
    onUpdate={[Function]}
    orientation="left"
    removeBtn={false}
    removeBtnLabel={null}
    syncCacheToGroup={[Function]}
    values={Object {}}
  >
    <div
      className="orizzonte__group-description"
      key="orizzonte-description"
    >
      Description for this group
    </div>
    <OrizzonteSelect
      disabled={false}
      fieldName="testSelect"
      information={null}
      label="Test select"
      notSetLabel={false}
      onUpdate={[Function]}
      options={
        Array [
          Object {
            "label": "Test value 1",
            "value": 1,
          },
          Object {
            "label": "Test value 2",
            "value": 2,
          },
        ]
      }
      value=""
    />
  </OrizzonteList>
</div>
`;

exports[`<Group /> should render group without clear / remove / done buttons, and higher min width 1`] = `
<div
  className="orizzonte__group orizzonte__group--shown orizzonte__group--empty"
  style={
    Object {
      "minWidth": "30px",
    }
  }
>
  <ForwardRef
    shown={true}
  >
    Test group
  </ForwardRef>
  <div
    className="orizzonte__group-btn-wrapper"
  >
    <button
      className="orizzonte__group-label"
      onClick={[Function]}
      type="button"
    >
      Test group
    </button>
    <OrizzonteGroupBtn
      hidden={true}
      onClick={[Function]}
    />
  </div>
  <OrizzonteList
    cache={Object {}}
    doneBtn={false}
    doneBtnLabel={null}
    isFilterGroup={true}
    minWidth={400}
    onApply={[Function]}
    onRemove={[Function]}
    onUpdate={[Function]}
    orientation="left"
    removeBtn={false}
    removeBtnLabel={null}
    syncCacheToGroup={[Function]}
    values={Object {}}
  >
    <OrizzonteSelect
      disabled={false}
      fieldName="testSelect"
      information={null}
      label="Test select"
      notSetLabel={false}
      onUpdate={[Function]}
      options={
        Array [
          Object {
            "label": "Test value 1",
            "value": 1,
          },
          Object {
            "label": "Test value 2",
            "value": 2,
          },
        ]
      }
      value=""
    />
  </OrizzonteList>
</div>
`;

exports[`<Group /> should render nothing when the group is not included 1`] = `""`;
